File: daily/Linked List/92-Reverse-linked-list-ii.go
package daily

func reverseBetween(head *ListNode, left int, right int) *ListNode {
	if head == nil || left == right {
		return head
	}

	dummy := &ListNode{0, head}
	before := dummy

	for i := 0; i < left-1; i++ {
		before = before.Next
	}

	curr := before.Next
	var prev *ListNode
	for i := 0; i < right-left+1; i++ {
		nextNode := curr.Next
		curr.Next = prev
		prev = curr
		curr = nextNode
	}
	before.Next.Next = curr
	before.Next = prev
	return dummy.Next
}

File: daily/Linked List/92-Reverse-linked-list-ii.py
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        before = dummy

        for _ in range(left - 1):
            before = before.next
        
        curr = before.next
        prev = None
        for _ in range(right - left + 1):
            nextNode = curr.next
            curr.next = prev
            prev = curr
            curr = nextNode
        
        before.next.next = curr
        before.next = prev
        return dummy.next